# Mexico Real Estate Analysis
## Data Cleaning & Preparation

### Research Question

> Are property prices in Mexico more strongly associated with
> property size or location?

### Objective

This project analyzes residential property listings in Mexico
to investigate the factors associated with property prices.

The first stage focuses on preparing the raw data for reliable
downstream analysis.

### This Notebook

In this notebook, I will:

1. Inspect the raw datasets
2. Identify data-quality and structural issues
3. Clean and standardize the datasets
4. Validate the cleaned data
5. Combine the datasets into one analytical dataset
6. Generate initial descriptive statistics

The cleaned dataset produced here will be used in subsequent
analysis notebooks.

In [1032]:
# Data manipulation
import pandas as pd
import numpy as np

## 1. Project Workflow

The data preparation process follows a simple workflow:

**Inspect → Diagnose → Clean → Validate → Combine → Summarize**

The goal is to understand the structure of the raw data before
making any transformations.

## 2. Load the Raw Data

The property listings are provided in three separate CSV files.

Each file contains information about residential properties, but the
datasets are not completely consistent in how the variables are
stored.

I will load each dataset separately first. This allows the structure
and quality of each source to be examined before any cleaning or
combination takes place.

In [1033]:
# Define paths to the raw datasets
path_1 = "../data/mexico-real-estate-1.csv"
path_2 = "../data/mexico-real-estate-2.csv"
path_3 = "../data/mexico-real-estate-3.csv"

In [1034]:
# Load the raw datasets
df1 = pd.read_csv(path_1)
df2 = pd.read_csv(path_2)
df3 = pd.read_csv(path_3)

The three datasets are kept as separate DataFrames during the
initial inspection stage.

This makes it possible to identify differences between the sources
before standardizing their structure.

## 3. Inspect the Raw Data

Before cleaning the data, I first need to understand what I have.

The initial inspection focuses on five questions:

1. How many observations and variables are present?
2. What columns are available?
3. Are the variables stored using appropriate data types?
4. Are there missing values?
5. Do the three datasets have a consistent structure?

This inspection helps identify problems before deciding how they
should be handled.

### 3.1 Dataset 1

I begin by inspecting the first dataset using a consistent set of
pandas inspection methods.

In [1035]:
# Display the first five rows
df1.head()

,property_type,state,lat,lon,area_m2,price_usd
0,house,Estado de México,19.560181,-99.233528,150.0,"$67,965.56"
1,house,Nuevo León,25.688436,-100.198807,186.0,"$63,223.78"
2,apartment,Guerrero,16.767704,-99.764383,82.0,"$84,298.37"
3,apartment,Guerrero,16.829782,-99.911012,150.0,"$94,308.80"
4,house,Veracruz de Ignacio de la Llave,NaN,NaN,175.0,"$94,835.67"


In [1036]:
# Check the number of rows and columns
df1.shape

(700, 6)

In [1037]:
# Inspect column types, non-null values, and memory usage
df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   property_type  700 non-null    str    
 1   state          700 non-null    str    
 2   lat            583 non-null    float64
 3   lon            583 non-null    float64
 4   area_m2        700 non-null    float64
 5   price_usd      700 non-null    str    
dtypes: float64(3), str(3)
memory usage: 32.9 KB


In [1038]:
# Check the data type of each column
df1.dtypes

property_type        str
state                str
lat              float64
lon              float64
area_m2          float64
price_usd            str
dtype: object

In [1039]:
# Count missing values in each column
df1.isnull().sum()

property_type      0
state              0
lat              117
lon              117
area_m2            0
price_usd          0
dtype: int64

### 3.2 Dataset 2

The same inspection process is applied to the second dataset.
Using the same checks makes it easier to compare the sources and
identify structural differences.

In [1040]:
# Display the first five rows
df2.head()

,property_type,state,lat,lon,area_m2,price_mxn
0,apartment,Nuevo León,25.721081,-100.345581,72.0,1300000.0
1,apartment,Puebla,NaN,NaN,190.0,2500000.0
2,house,Morelos,23.634501,-102.552788,360.0,5300000.0
3,house,Morelos,NaN,NaN,76.0,820000.0
4,house,Puebla,NaN,NaN,200.0,1100000.0


In [1041]:
# Check the number of rows and columns
df2.shape

(700, 6)

In [1042]:
# Inspect column types, non-null values and memory usage
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   property_type  700 non-null    str    
 1   state          700 non-null    str    
 2   lat            571 non-null    float64
 3   lon            571 non-null    float64
 4   area_m2        700 non-null    float64
 5   price_mxn      700 non-null    float64
dtypes: float64(4), str(2)
memory usage: 32.9 KB


In [1043]:
# Check the data type of each column
df2.dtypes

property_type        str
state                str
lat              float64
lon              float64
area_m2          float64
price_mxn        float64
dtype: object

In [1044]:
# Check missing values in each column
df2.isnull().sum()

property_type      0
state              0
lat              129
lon              129
area_m2            0
price_mxn          0
dtype: int64

### 3.3 Dataset 3

Finally, I inspect the third dataset using the same procedure.

In [1045]:
# Display the first five rows
df3.head()

,property_type,place_with_parent_names,lat-lon,area_m2,price_usd
0,apartment,|México|Distrito Federal|Gustavo A. Madero|Acu...,"19.52589,-99.151703",71.0,48550.59
1,house,|México|Estado de México|Toluca|Metepec|,"19.2640539,-99.5727534",233.0,168636.73
2,house,|México|Estado de México|Toluca|Toluca de Lerd...,"19.268629,-99.671722",300.0,86932.69
3,house,|México|Morelos|Temixco|Burgos Bugambilias|,NaN,275.0,263432.41
4,apartment,|México|Veracruz de Ignacio de la Llave|Veracruz|,"19.511938,-96.871956",84.0,68508.67


In [1046]:
# Check the number of rows and columns
df3.shape

(700, 5)

In [1047]:
# Check column types, non-null values and memory usage
df3.info()

<class 'pandas.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   property_type            700 non-null    str    
 1   place_with_parent_names  700 non-null    str    
 2   lat-lon                  582 non-null    str    
 3   area_m2                  700 non-null    float64
 4   price_usd                700 non-null    float64
dtypes: float64(2), str(3)
memory usage: 27.5 KB


In [1048]:
# Check the data type for each column
df3.dtypes

property_type                  str
place_with_parent_names        str
lat-lon                        str
area_m2                    float64
price_usd                  float64
dtype: object

In [1049]:
# Count missing values in each column
df3.isnull().sum()

property_type                0
place_with_parent_names      0
lat-lon                    118
area_m2                      0
price_usd                    0
dtype: int64

## Compare Dataset Structures
The three datasets represent the same general subject—Mexican
property listings—but they may differ in their column names,
number of variables, and representation of individual values.

Before combining them, I compare their schemas to determine what
must be standardized.

In [1050]:
# Display the column names for eeach dataset
print("Dataset 1:", df1.columns.tolist())
print("Dataset 2:", df2.columns.tolist())
print("Dataset 3:", df3.columns.tolist())

Dataset 1: ['property_type', 'state', 'lat', 'lon', 'area_m2', 'price_usd']
Dataset 2: ['property_type', 'state', 'lat', 'lon', 'area_m2', 'price_mxn']
Dataset 3: ['property_type', 'place_with_parent_names', 'lat-lon', 'area_m2', 'price_usd']


### Initial Findings

The inspection reveals several differences between the data sources:

- Dataset 1 contains `lat` and `lon` as separate columns.
- Dataset 2 stores the property price in Mexican pesos (`price_mxn`).
- Dataset 3 combines latitude and longitude into a single `lat-lon`
  column.
- Dataset 3 stores location information inside
  `place_with_parent_names`.
- Some datasets contain missing values that must be addressed.
- The representation of `price_usd` is not consistent across all
  sources.

These differences mean that the datasets cannot be combined
directly. Each source must first be cleaned and standardized to a
common schema.

## 5. Data Quality Assessment

The inspection stage identified several data-quality and structural
issues that must be resolved before analysis.

The main issues are:

| Dataset | Issue | Required Action |
|---|---|---|
| Dataset 1 | Missing latitude and longitude | Remove incomplete coordinate records |
| Dataset 1 | `price_usd` stored as text | Remove `$` and commas, then convert to numeric |
| Dataset 2 | Price stored in MXN | Convert price to USD |
| Dataset 2 | Missing values | Remove incomplete records |
| Dataset 3 | Latitude and longitude combined | Split into separate columns |
| Dataset 3 | State embedded in location string | Extract the state |
| Dataset 3 | Composite source columns | Remove after extracting required variables |

The objective is not simply to remove problematic data. Each
transformation should produce a consistent structure that supports
the research question.

## 6. Clean Dataset 1

The first dataset already contains the variables needed for the
analysis, but two data-quality issues need to be addressed:

- `price_usd` is stored as text rather than a numeric value.
- Some properties have missing latitude or longitude values.

Because location is one of the factors in the research question,
records without coordinates cannot be used for location-based
analysis. These incomplete records will therefore be removed.

The cleaning process will:

1. Remove rows with missing values
2. Clean the `price_usd` values
3. Convert `price_usd` from text to numeric

In [1051]:
# Clean and standardize Dataset 1
df1 = (
    pd.read_csv(path_1)
    .dropna()
    .assign(
        price_usd=lambda x:(
            x["price_usd"]
            .str.replace("$", "", regex=False)
            .str.replace(",", "", regex=False)
            .astype(float)
        )
    )

)

### Cleaning the Price Variable

The `price_usd` column contains values such as:

`$125,000`

These values are stored as strings because of the dollar sign and
thousands separator.

To make the variable suitable for numerical analysis, I:

1. Remove the `$` symbol
2. Remove the comma
3. Convert the resulting string to `float`

The `regex=False` argument ensures that `$` is interpreted as a
literal character rather than a regular-expression symbol.

In [1052]:
# Verify the cleaned structure
df1.head()

,property_type,state,lat,lon,area_m2,price_usd
0,house,Estado de México,19.560181,-99.233528,150.0,67965.56
1,house,Nuevo León,25.688436,-100.198807,186.0,63223.78
2,apartment,Guerrero,16.767704,-99.764383,82.0,84298.37
3,apartment,Guerrero,16.829782,-99.911012,150.0,94308.80
5,house,Yucatán,21.052583,-89.538639,205.0,105191.37


In [1053]:
# Check the resulting dimensions
df1.shape

(583, 6)

In [1054]:
# Confirm the data types
df1.dtypes

property_type        str
state                str
lat              float64
lon              float64
area_m2          float64
price_usd        float64
dtype: object

### Dataset 1 Result

After cleaning, Dataset 1 contains **583 observations and 6
variables**.

The dataset now has:

- separate latitude and longitude columns
- numeric `price_usd` values
- no missing values in the retained records
- a structure suitable for combination with the other datasets

## 7. Clean Dataset 2

Dataset 2 has the same general property information as Dataset 1,
but its price variable is stored in Mexican pesos (`price_mxn`)
instead of US dollars.

To combine the datasets consistently, prices must be converted to
the same currency.

For this project, the conversion uses:

> **19 MXN = 1 USD**

The cleaning process will:

1. Convert `price_mxn` to USD
2. Create a standardized `price_usd` column
3. Remove the original `price_mxn` column
4. Remove incomplete records

In [1055]:
df2 = (
    pd.read_csv(path_2)
    .assign(
        price_usd = lambda x: x["price_mxn"].div(19)
    )
    .drop(columns=["price_mxn"])
    .dropna()
)




### Currency Conversion

The conversion is performed using pandas' `.div()` method:

```python
x["price_mxn"].div(19)

In [1056]:

# Preview the cleaned dataset
df2.head()

,property_type,state,lat,lon,area_m2,price_usd
0,apartment,Nuevo León,25.721081,-100.345581,72.0,68421.052632
2,house,Morelos,23.634501,-102.552788,360.0,278947.368421
6,apartment,Estado de México,19.272040,-99.572013,85.0,65789.473684
7,house,San Luis Potosí,22.138882,-100.996510,158.0,111578.947368
8,apartment,Distrito Federal,19.394558,-99.129707,65.0,39904.736842


In [1057]:
# Check the dimensions
df2.shape

(571, 6)

In [1058]:
# Confirm the standardized column types
df2.dtypes

property_type        str
state                str
lat              float64
lon              float64
area_m2          float64
price_usd        float64
dtype: object

### Dataset 2 Result

After cleaning, Dataset 2 contains **571 observations and 6
variables**.

The price variable is now expressed in US dollars, and the original
`price_mxn` column has been removed.

Dataset 2 is therefore ready to be standardized with the remaining
sources.

## 8. Clean Dataset 3

Dataset 3 requires more structural transformation than the first
two datasets.

Two columns contain multiple pieces of information:

- `lat-lon` contains both latitude and longitude in one value.
- `place_with_parent_names` contains hierarchical location
  information, including the country, state, and city.

These composite fields do not follow tidy-data principles because
multiple variables are stored together.

The cleaning process will:

1. Remove incomplete records
2. Split `lat-lon` into `lat` and `lon`
3. Extract the state from `place_with_parent_names`
4. Remove the original composite columns

In [1059]:
# Clean and standardize dataset 3
df3 = (
    pd.read_csv(path_3)
    .dropna()
    .assign(
        # Extract lat from lat-lon column
        lat = lambda x: (
            x["lat-lon"]
            .str.split(",", expand=True)[0]
            .astype(float)
        ),
        # Extract lon from lat-lon column
        lon = lambda x: (
            x["lat-lon"]
            .str.split(",", expand=True)[1]
            .astype(float)
            
        ),
        # Extract state from place_with_parent_names column
        state = lambda x: (
            x["place_with_parent_names"]
            .str.split("|", expand=True)[2]
        )
        

     )
    .drop(columns=["place_with_parent_names", "lat-lon"])
)

### Extracting Latitude and Longitude

The original `lat-lon` column stores two variables in one cell.

For example:

`19.4326,-99.1332`

The string is split at the comma:

```python
.str.split(",", expand=True)

In [1060]:
# Preview the cleaned dataset
df3.head()

,property_type,area_m2,price_usd,lat,lon,state
0,apartment,71.0,48550.59,19.525890,-99.151703,Distrito Federal
1,house,233.0,168636.73,19.264054,-99.572753,Estado de México
2,house,300.0,86932.69,19.268629,-99.671722,Estado de México
4,apartment,84.0,68508.67,19.511938,-96.871956,Veracruz de Ignacio de la Llave
5,house,175.0,102763.00,20.689157,-103.366728,Jalisco


### Extracting the State

The `place_with_parent_names` column contains hierarchical
location information such as:

`|Mexico|Jalisco|Guadalajara|`

Splitting the string at `|` produces:

| Index | Value |
|---:|---|
| 0 | empty |
| 1 | Mexico |
| 2 | Jalisco |
| 3 | Guadalajara |

The state is therefore extracted using index `2`.

In [1061]:
# Check the dimensions after cleaning
df3.shape

(582, 6)

In [1062]:
# Confirm the standardized structure
df3.dtypes

property_type        str
area_m2          float64
price_usd        float64
lat              float64
lon              float64
state                str
dtype: object

### Dataset 3 Result

After cleaning, Dataset 3 contains **582 observations and 6
variables**.

The composite location fields have been transformed into separate
variables, giving the dataset the same analytical structure as
Datasets 1 and 2.

## 9. Validate the Cleaned Datasets

Before combining the three datasets, I need to confirm that they
share the same analytical structure.

All three datasets should contain the same six variables:

- `property_type`
- `state`
- `lat`
- `lon`
- `area_m2`
- `price_usd`

This validation step reduces the risk of introducing incorrectly
aligned columns when the datasets are combined.

In [1063]:
# Display the columns in each cleaned dataset
print("Dataset 1:", df1.columns.tolist())
print("Dataset 2:", df2.columns.tolist())
print("Dataset 3:", df3.columns.tolist())

Dataset 1: ['property_type', 'state', 'lat', 'lon', 'area_m2', 'price_usd']
Dataset 2: ['property_type', 'state', 'lat', 'lon', 'area_m2', 'price_usd']
Dataset 3: ['property_type', 'area_m2', 'price_usd', 'lat', 'lon', 'state']


### Check Column Names

The column names should be identical across all three datasets.

However, matching names alone are not enough. The order of the
columns should also be consistent before combining the data.

In [1064]:
# Check whether all datasets have the same columns in the same order
columns_match = (
    df1.columns.tolist()
    == df2.columns.tolist()
    == df3.columns.tolist()
)

columns_match

False

### Standardize Column Structure

Before combining the cleaned datasets, I verify that they contain the same columns in the same order. This is important because the datasets come from different sources and may have different column arrangements.

Although all three datasets contain the same six variables, Dataset 3 has them in a different order. I therefore standardize the column order across all datasets before concatenation.

In [1065]:
# Standardize column order in dataset 3
df3 = df3[
    ["property_type", "state", "lat", "lon", "area_m2", "price_usd"]
]

In [1066]:
# Check whether all datasets have the same columns in the same order
columns_match = (
    df1.columns.tolist()
    == df2.columns.tolist()
    == df3.columns.tolist()
)

columns_match

True

A result of `True` confirms that the three cleaned datasets have
matching column names in the same order.

This means they can be safely stacked row-wise using `pd.concat()`.

## 10. Combine the Datasets

The three cleaned datasets now have a consistent schema, so they can
be combined into a single analytical dataset.

Because each dataset contains different property listings with the
same variables, the appropriate operation is to **stack the rows**
using `pd.concat()`.

The `ignore_index=True` argument creates a new sequential index for
the combined dataset.

In [1067]:
# Combine the three cleaned datasets
df = pd.concat(
    [df1, df2, df3],
    ignore_index=True
)

### Why `pd.concat()`?

`pd.concat()` is appropriate because the datasets contain the same
variables but different observations.

Conceptually:

```text
Dataset 1
────────────
Property 1
Property 2
Property 3
     ↓
Dataset 2
────────────
Property 4
Property 5
Property 6
     ↓
Dataset 3
────────────
Property 7
Property 8
Property 9

In [1068]:
# Inspect the combined dataset
df.head()

,property_type,state,lat,lon,area_m2,price_usd
0,house,Estado de México,19.560181,-99.233528,150.0,67965.56
1,house,Nuevo León,25.688436,-100.198807,186.0,63223.78
2,apartment,Guerrero,16.767704,-99.764383,82.0,84298.37
3,apartment,Guerrero,16.829782,-99.911012,150.0,94308.80
4,house,Yucatán,21.052583,-89.538639,205.0,105191.37


In [1069]:
# Check the dimensions of the combined dataset
df.shape

(1736, 6)

### Expected Result

The combined dataset should contain:

- **1,736 observations**
- **6 variables**

The row count should equal the sum of the cleaned datasets:

\[
583 + 571 + 582 = 1,736
\]

No observations should be lost during concatenation.

## 11. Validate the Combined Dataset

After combining the datasets, I perform a final structural check.

The validation confirms that:

- The expected number of observations is present
- Mll six variables are available
- Missing values have not been reintroduced
- The index is sequential
- The variables have appropriate data types

In [1070]:
# Check the combined dataset structure
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1736 entries, 0 to 1735
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   property_type  1736 non-null   str    
 1   state          1736 non-null   str    
 2   lat            1736 non-null   float64
 3   lon            1736 non-null   float64
 4   area_m2        1736 non-null   float64
 5   price_usd      1736 non-null   float64
dtypes: float64(4), str(2)
memory usage: 81.5 KB


In [1071]:
# Check for missing values
df.isnull().sum()

property_type    0
state            0
lat              0
lon              0
area_m2          0
price_usd        0
dtype: int64

In [1072]:
# Check the index
df.index

RangeIndex(start=0, stop=1736, step=1)

In [1073]:
# Confirm the final columns
df.columns.tolist()

['property_type', 'state', 'lat', 'lon', 'area_m2', 'price_usd']

### Validation Result

The combined dataset contains **1,736 property listings and 6
variables**.

The datasets have been successfully standardized and combined
without introducing additional missing values.

The resulting DataFrame is now ready for descriptive analysis.

## 12. Initial Descriptive Statistics

Before investigating relationships between property characteristics
and price, I first examine the numerical variables.

The `price_usd` and `area_m2` variables are summarized using
`.describe()`.

This provides measures of central tendency, variability, and the
range of observed values.

In [1074]:
# Generate descriptive statistics for price and property area
summary = df[["price_usd", "area_m2"]].describe()

summary

,price_usd,area_m2
count,1736.000000,1736.000000
mean,115331.980800,170.261521
std,65426.173793,80.594539
min,33157.894737,60.000000
25%,65789.473684,101.750000
50%,99262.132105,156.000000
75%,150846.665000,220.000000
max,326733.660000,385.000000


### Interpreting the Summary

The descriptive statistics provide an initial understanding of the
property market represented in the dataset.

The mean property price is approximately **$115,332**, while the
median is approximately **$99,262**.

Because the mean is higher than the median, the price distribution
shows evidence of **right skewness**. A relatively small number of
higher-priced properties are likely pulling the mean upward.

For this reason, the median can provide a more representative
measure of a typical property price than the mean.

Property size shows a similar need for distributional analysis,
with a mean of approximately **170 m²** and a median of **156 m²**.

In [1075]:
# Compare the mean and median of the main numerical variables
df[["price_usd", "area_m2"]].agg(["mean", "median"])

,price_usd,area_m2
mean,115331.980800,170.261521
median,99262.132105,156.000000


## 13. Save the Clean Dataset

The cleaned and validated dataset is saved as a new CSV file.

Keeping the cleaned dataset separate from the raw source files
preserves the original data while creating a reusable input for
future analysis.

In [ ]:
# Save the cleaned dataset
output_path="../data/mexico-real-estate-combined-clean.csv"
df.to_csv(output_path, index=False)

: 

## Conclusion

The data preparation stage is now complete.

Three raw property datasets were inspected, cleaned, standardized,
validated, and combined into a single analytical dataset containing
**1,736 property listings and 6 variables**.

The final dataset contains consistent representations of property
type, location, area, and price, making it suitable for downstream
analysis.

The next stage of the project will investigate the research question
more directly by examining the relationship between **property
size, location, and price**.